### PR TITLE
Fix SOTA to not reboot twice on EXT4 system

### PR DESCRIPTION
**Pull request by tsirlapu**
RTC 530960

Fix SOTA snapshot conditionali statements

_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
_Provide a 1-2 line brief overview of the changes submitted through the Pull Request..._
Fix SOTA snapshot conditionali statements

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): **\<URL to be filled>**
https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=530960&oslc_config.context=https://rtc.intel.com/gc/configuration/4491
- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [x] Changelogs are updated (or N/A if not customer visible)
- [x] inbm/log_changes.txt and inbm-vision/log_changes.txt are updated for potentially Validation-breaking log changes (or N/A if none)

### CODE MAINTAINABILITY
- [ ] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [ ] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [ ] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):
- [ ] Run 'go fmt' or format-python.sh as applicable.
- [ ] New/modified methods and functions should have type annotations on signatures as applicable
- [ ] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**
- [ ] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

### CODE REVIEW

**Comment by nex-maximus** [comment]
## Code Review Analysis

- 🎯 **Main theme:** The PR addresses a specific issue where a system would reboot twice during a SOTA (Software Over The Air) update on EXT4 filesystems. The changes involve modifying conditional checks to prevent unnecessary rollbacks and reboots when the snapshot number is "0".
- 📝 **PR summary:** This PR introduces changes to the snapshot handling logic in the SOTA update process. It modifies the conditions under which a rollback and reboot are triggered, specifically avoiding these actions when the snapshot number is "0", which indicates that no snapshot was taken.
- 📌 **Type of PR:** Bug fix
- 🏅 **Score:** 85
- 🧪 **Relevant tests added:** No
- ⏱️ **Estimated effort to review [1-5]:** 2, because the PR is focused on a specific issue with a clear description, and the changes are limited to a few lines of code in a single module.
- 🔒 **Security concerns:** No
## Code Review Feedback



<strong>💡 General suggestions:</strong> The PR seems to address the issue described, and the changes are clear and concise. However, it is important to ensure that the snapshot number "0" is indeed the correct indicator for no snapshot taken and that there are no edge cases where "0" could be a valid snapshot number. Additionally, it would be beneficial to add relevant unit tests to cover these changes and ensure that the new logic behaves as expected in different scenarios.


<details> <summary><strong>✨ Usage tips:</strong></summary><hr> 

> Tag me in a comment '@nex-maximus' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 

</details>

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py [hunk 1]
```diff
@@ -171,7 +171,7 @@ def _rollback_and_delete_snap(self) -> None:
         a.) After reboot by SOTA, and diagnostic reports bad report for system health
         """
         logger.debug("")
-        if self.snap_num:
+        if self.snap_num and self.snap_num is not "0":
```
Comment (nex-maximus) [inline inbm/dispatcher-agent/dispatcher/sota/snapshot.py:174]: <h3>$\color{#FFA500}Improvements: Code Maintainability$</h3>

Consider using `!=` instead of `is not` for comparing `self.snap_num` with the string "0" to avoid potential issues with string identity checks.

Comment (nex-maximus) [inline inbm/dispatcher-agent/dispatcher/sota/snapshot.py:174]: <h3>$\color{#FFA500}Improvements: Code Maintainability$</h3>

Add a comment explaining why the snapshot number "0" is treated as a special case, for future maintainability.

Comment (nex-maximus) [inline inbm/dispatcher-agent/dispatcher/sota/snapshot.py:174]: <h3>$\color{#FFA500}Improvements: Code Robustness$</h3>

Ensure that `self.snap_num` is always a string before performing string comparisons, or convert it explicitly to avoid type-related bugs.

File: inbm/Changelog.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Fixed
  - RTC 538468 - paho-mqtt upgrade broke cloudadapter's mqtt connections. Fixed proxy setting code to not override all sockets with proxy as paho-mqtt 1.6.0 relies on listening/connecting to localhost to set up sockets, and this doesn't work with a global proxy on all sockets.
  - RTC 538549 - improved errors when unable to fetch from URLs. For example, if INBM receives a "404 Not Found" it will return this as part of its error instead of simply returning a generic error message about being unable to fetch the URL.
+ - RTC 530960 - Fix SOTA to not reboot twice on EXT4 system
```
Comment (nex-maximus) [inline inbm/Changelog.md:14]: <h3>$\color{#FFA500}Improvements: Code Maintainability$</h3>

Provide more detail in the Changelog about the conditions under which the bug occurs and the impact of the fix.

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py [hunk 3]
```diff
@@ -171,7 +171,7 @@ def _rollback_and_delete_snap(self) -> None:
...
             self._dispatcher_broker.telemetry("SOTA attempting rollback")
             rc, err = self.trtl.sota_rollback(self.snap_num)
```
Comment (nex-maximus) [inline inbm/dispatcher-agent/dispatcher/sota/snapshot.py:176]: **<h2>$\color{#FFA500}Improvements: Code Maintainability$</h2>**
**Suggestion:** Replace the identity check 'is not' with the equality operator '!=' when comparing strings.
**NOTE:** Ensure that the suggested code is comprehensive. This serves as an example of the code you could use. The user should verify it after accepting the suggestion.
```suggestion
        if self.snap_num and self.snap_num != "0":
            self._dispatcher_broker.telemetry("SOTA attempting rollback")
            rc, err = self.trtl.sota_rollback(self.snap_num)
```

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py [hunk 4]
```diff
@@ -224,7 +224,7 @@ def recover(
         @param time_to_wait_before_reboot: If we are rebooting, wait this many seconds first.
         """
         dispatcher_state.clear_dispatcher_state()
-        if self.snap_num:
+        if self.snap_num and self.snap_num is not "0":
             self._rollback_and_delete_snap()
             logger.debug("Rebooting to recover from failed SOTA...")
             rebooter.reboot()
```
Comment (nex-maximus) [inline inbm/dispatcher-agent/dispatcher/sota/snapshot.py:230]: **<h2>$\color{#FFA500}Improvements: Code Maintainability$</h2>**
**Suggestion:** Ensure that the 'snap_num' is compared as a string since it is expected to be a string value.
**NOTE:** Ensure that the suggested code is comprehensive. This serves as an example of the code you could use. The user should verify it after accepting the suggestion.
```suggestion
        if self.snap_num and self.snap_num != "0":
            self._rollback_and_delete_snap()
            logger.debug("Rebooting to recover from failed SOTA...")
            rebooter.reboot()
```

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py [hunk 5]
```diff
@@ -245,7 +245,7 @@ def revert(self, rebooter: Rebooter, time_to_wait_before_reboot: int) -> None:
         """
         logger.debug("")
         dispatcher_state.clear_dispatcher_state()
-        if self.snap_num:
+        if self.snap_num and self.snap_num is not "0":
             self._rollback_and_delete_snap()
             time.sleep(time_to_wait_before_reboot)
             rebooter.reboot()
```
Comment (nex-maximus) [inline inbm/dispatcher-agent/dispatcher/sota/snapshot.py:251]: **<h2>$\color{#FFA500}Improvements: Code Maintainability$</h2>**
**Suggestion:** Use the equality operator '!=' for string comparison to maintain consistency and readability.
**NOTE:** Ensure that the suggested code is comprehensive. This serves as an example of the code you could use. The user should verify it after accepting the suggestion.
```suggestion
        if self.snap_num and self.snap_num != "0":
            self._rollback_and_delete_snap()
            time.sleep(time_to_wait_before_reboot)
            rebooter.reboot()
```

File: inbm/Changelog.md (hunk 2 above)
Comment (gblewis1) [reply inbm/Changelog.md]: @tsirlapu I see this is marked with a thumbs up but I don't see that it's been implemented in Changelog

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py (hunk 3 above)
Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/snapshot.py]: Agreed. Using  '!=' directly for string comparison is more idiomatic and clearer

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py (hunk 4 above)
Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/snapshot.py]: Agreed. Using  '!=' directly for string comparison is more idiomatic and clearer

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py (hunk 5 above)
Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/snapshot.py]: Agreed. Using  '!=' directly for string comparison is more idiomatic and clearer

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py (hunk 1 above)
Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/snapshot.py]: Yes already it is declared.

Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/sota/snapshot.py]: @tsirlapu "is not" and "!=" do different things in Python--idiomatic would be if they did the same thing but one was more familiar. see https://www.geeksforgeeks.org/difference-between-and-is-not-operator-in-python/

File: inbm/Changelog.md (hunk 2 above)
Comment (tsirlapu) [reply inbm/Changelog.md]: @gblewis1 I changed it to - Fix SOTA snapshot conditions to not reboot twice on EXT4 system

**Review by gblewis1 — APPROVED** [review]

**Comment by nex-maximus** [comment]
**<h2>$\color{#FFA500}CODE \space REVIEW \space STARTED$</h2>**

**Comment by nex-maximus** [comment]
Preparing review...

**Comment by nex-maximus** [comment]
## Incremental PR Review

- **Review for commits since previous PR-Agent review:** Starting from commit https://github.com/intel/intel-inb-manageability/pull/481/commits/bae588093eeaec93bb78e9116ab3a40ea92ae161  
_(Merge branch 'sota\_snap0' of https://github.com/intel/intel-inb-manageability into sota\_snap0)_
## Code Review Analysis

- **PR summary:** This PR addresses an issue where systems with EXT4 filesystems would reboot twice during the SOTA (Software Over The Air) update process. It corrects conditional statements related to snapshot management to ensure that unnecessary reboots are avoided, enhancing the update process's efficiency and reliability.
- **PR Category:** Bug fix
- **Security concerns:** No
## Code Review Feedback



<strong> General suggestions:</strong> The PR successfully addresses the critical issue of avoiding unnecessary reboots during the SOTA process on EXT4 systems by fixing conditional checks. It's essential to ensure that such changes are thoroughly tested across different system configurations to prevent any unforeseen issues. Additionally, considering the importance of sustainability mentioned in the PR description, it would be beneficial to include inline comments explaining the rationale behind key changes for future maintainability.


<details> <summary><strong> User guidelines:</strong></summary><hr> 

> Tag me in a comment '@nex-maximus' and add one of the following commands:
> **/review [-all]**: Request a review of your Pull Request. By default, the agent considers changes since the last review For including all the files in PR, which only include the '-all' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 

</details>

**Comment by nex-maximus** [comment]
Preparing code suggestions...

**Comment by nex-maximus** [comment]
**<h2>$\color{#FFA500}CODE \space REVIEW \space ENDED$</h2>**
> If you wish to include any dependent files which are not in the PR, add the details in:
            1. .github/dependent_files.json
            (or)
            2. pass it via the comment in this format -> /review -all --dependent_files=file1.py, scr2/file2.py
